Add BPF map sizes to output of `cilium status --verbose`

tklauser opened the pull request:
When the dynamic BPF map sizing feature is used, for debugging, it would be useful to know the actual map sizes without asking users to run `bpftool map show pinned ...` or grepping logs of the agent.
    
Thus, report the BPF map sizes in and the dynamic size ratio in `cilium status --verbose`.

Note 1: The first commit removes a trailing `"` from an API description and the second commit does some preparatory refactoring to simplify add new verbosity options `pkg/client.FormatStatusResponse`. See commit messages for details.

Note 2: Not sure whether this should be backported to 1.8 given the refactoring. Though it might be useful to have the BPF map size status information to debug user issues. Comments welcome.

Fixes: #12643

```release-note
Add BPF map sizes to output of `cilium status --verbose`
```

tklauser committed:
api: remove trailing quotation mark in BPFMapEntry description

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

brb commented on pkg/client/client.go:
```diff
@@ -240,13 +240,30 @@ func numReadyClusters(clustermesh *models.ClusterMeshStatus) int {
+// FormatStatusResponse writes a StatusResponse as a string to the writer. The bit mask sd controls
+// whether a additional details are printed about a certain aspect of the status. In case there are
+// errors, some details may be printed regardless of the value of sd.
+func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetails) {
```
Nit: have you considered using struct for StatusDetails instead?

tklauser replied:
No haven't considered it so far, thanks for the suggestion. A `StatusDetails struct` of `bool`s would make the flag parsing in `cilium/cmd` nicer. Will change.

brb commented on pkg/client/client.go:
```diff
@@ -500,4 +503,18 @@ func FormatStatusResponse(w io.Writer, sr *models.StatusResponse, sd StatusDetai
+		fmt.Fprintf(w, "  CT any\t%d\n", sr.BpfMaps.CtAnyMapSize)
+		fmt.Fprintf(w, "  NAT\t%d\n", sr.BpfMaps.NatMapSize)
+		fmt.Fprintf(w, "  Neigh\t%d\n", sr.BpfMaps.NeighMapSize)
+		fmt.Fprintf(w, "  SockRev\t%d\n", sr.BpfMaps.SockRevMapSize)
```
Nit: maybe we should include all maps managed by Cilium?

tklauser replied:
Yes, makes sense.

brb reviewed (approved):
LGTM, thanks! Two nits/questions.

tklauser committed:
client: refactor FormatStatusResponse to take struct for additional details

Adding a bool var for each additional details to print in the status
makes requires changing the function and all its callers and additionaly
makes certain invocations of the function hard to read without looking
at its signature and godoc comment due to the list of bools.

Refactor FormatStatusResponse to take a struct of bools with the
additional details to print. This e.g. makes the invocation of a caller
not interested in any details more readable as:

    client.FormatStatusResponse(w, sr, client.StatusNoDetails)

compared to the previous version:

    client.FormatStatusResponse(w, sr, false, false, false, false, false)

Adding a new option for additional details now only requires adding an
additional bool member and its handling in FormatStatusResponse and
StatusAllDetails. Only callers interested in this particular detail need
to know about this flag and set it, existing callers don't need to be
changed.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser committed:
maps/ipmasq: export MaxEntries

This is needed to report the ipmasq map size to report in `cilium status
--verbose` in a successive commit.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

joestringer commented on api/v1/openapi.yaml:
```diff
@@ -2399,6 +2402,28 @@ definitions:
       last-error:
         description: Last error seen while performing desired action
         type: string
+  BPFMapSizes:
```
I wonder from an API perspective if this should be "BPF Maps" at the top level, then nested within, objects that represent maps with a name and a size. Just in case we end up wanting to include any other information subsequently. It'd be slightly more generic than just "Map sizes".

tklauser replied:
Agree. I changed this to just `BPFMapStatus` at the top level already. Will also change the lower level to be more generic and allow for other information than just size in the future.

tklauser replied:
Changed this to an array of `BPFMapProperties` which is filled by the agent. The map names are no longer hard-coded in the API. This allows to add/remove maps without changing the API. Currently `BPFMapProperties` contains map name and size but may be extended in the future for additional use cases. PTAL.

joestringer commented on cilium-health/cmd/get.go:
```diff
@@ -47,7 +47,7 @@ var healthGetCmd = &cobra.Command{
 			fmt.Fprintf(w, "Node load:\t%s %s %s\n",
 				load.Last1min, load.Last5min, load.Last15min)
-			ciliumClient.FormatStatusResponse(w, sr.Cilium, false, false, false, false, false)
+			ciliumClient.FormatStatusResponse(w, sr.Cilium, ciliumClient.StatusNoDetails)
```
:tada:

joestringer commented on cilium/cmd/status.go:
```diff
@@ -58,7 +59,7 @@ func init() {
 	statusCmd.Flags().BoolVar(&allClusters, "all-clusters", false, "Show all clusters")
 	statusCmd.Flags().BoolVar(&brief, "brief", false, "Only print a one-line status message")
-	statusCmd.Flags().BoolVar(&verbose, "verbose", false, "Equivalent to --all-addresses --all-controllers --all-nodes --all-health")
+	statusCmd.Flags().BoolVar(&verbose, "verbose", false, "Equivalent to --all-addresses --all-controllers --all-nodes --all-health --all-redirects --all-clusters")
```
minor nit, now that there's a `StatusDetails` object, I wonder if we could do `StatusAllDetails.String()` and have that object print each of these. Then if we expand it any further, we should extend `StatusAllDetails()` too?

Not a big deal though, even if we forget to update this help message it doesn't matter so much.

tklauser replied:
Good point, will implement `StatusDetails.String()`.

tklauser replied:
I decided against adding `StatusDetails.String()` since it would move CLI specific flag names to `pkg/client` which is not only used by CLI. Maybe we could just adjust the flag usage description to say something like `Verbose output. Includes details about certain status aspects` and no longer specify the flags it implies.

joestringer reviewed (changes requested):
I like the tidy up of the status checks using a dedicated type.

One main comment below on the API, I think it may be worth making it slightly more generic to handle BPF maps. (Or we could plausibly even go one step further, create a BPF section with maps underneath then actual map objects under that with sizes).

tklauser committed:
maps/{eventsmap,signalmap}: pass map size as parameter to InitMap

Both the events and the signal map are sized to the number of possible
CPUs. Instead of twice querying the respective number, pass it as a
parameter to InitMap, so the number of possible CPUs can be determined
once before creating both maps.

Also export the map size as MaxEntries for both maps. This will be used
in a successive commit to report the map size in `cilium status
--verbose`.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

joestringer commented on daemon/cmd/status.go:
```diff
@@ -200,6 +210,82 @@ func (d *Daemon) getKubeProxyReplacementStatus() *models.KubeProxyReplacement {
 	}
 }
 
+func (d *Daemon) getBPFMapStatus() *models.BPFMapStatus {
```
We could plausibly have this iterate a structure held in the daemon, where that structure is populated via each map registering themselves, eg by passing the daemon into each map creation function which would receive an interface that represents the 'map manager' (with `RegisterMap(...)` method), then have the core bpf lib code automatically register the maps using their name / sizes at creation time. This would mean that we don't have to update this extra function whenever we add/remove maps, it would be automatically kept up-to-date by the core bpf map lib code.

But up to you whether to explore this now or later, it's not such a big problem that I think we need to solve this.

tklauser replied:
Nice, thanks for the suggestion! I like this approach and will look into implementing it in a follow-up PR.

joestringer reviewed (approved):
New approach looks great. I have one suggestion below that would allow us to reduce the dependency from the `daemon` package onto each map by autoregistering all maps with the daemon during initialization time (via dependency injection) but I don't think it's a big deal whether to implement that way or not.

brb commented on daemon/cmd/status.go:
```diff
@@ -200,6 +210,82 @@ func (d *Daemon) getKubeProxyReplacementStatus() *models.KubeProxyReplacement {
+			{
+				Name: "Tunnel",
+				Size: int64(tunnelmap.MaxEntries),
+			},
```
Although lbmap sizes are static, but for completion I'd suggest to add the following maps: `Service{4,6}MapV2`, `Backend{4,6}Map`, `RevNat{4,6}Map`.

tklauser replied:
Added, PTAL.

tklauser committed:
api, cli, daemon: add BPF map sizes to cilium status

When the dynamic BPF map sizing feature is used, for debugging, it would
be useful to know the actual map sizes without asking users to run
bpftool map show pinned ... or grepping logs of the agent.

Thus, report the BPF map sizes in and the dynamic size ratio in
`cilium status --verbose`.

Example output:

$ cilium status --verbose
[...]
BPF Maps:       dynamic sizing: on (ratio: 0.030000)
  Name                          Size
  Non-TCP connection tracking   217226
  TCP connection tracking       434452
  Endpoint policy               65535
  Events                        2
  IP cache                      512000
  IP masquerading agent         16384
  IPv4 fragmentation            8192
  IPv4 service                  65536
  IPv6 service                  65536
  IPv4 service backend          65536
  IPv6 service backend          65536
  IPv4 service reverse NAT      65536
  IPv6 service reverse NAT      65536
  Metrics                       1024
  NAT                           434452
  Neighbor table                434452
  Global policy                 16384
  Per endpoint policy           65536
  Session affinity              65536
  Signal                        2
  Sockmap                       65535
  Sock reverse NAT              217226
  Tunnel                        65536
[...]

Fixes #12643

Suggested-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

brb commented on daemon/cmd/status.go:
```diff
@@ -200,6 +210,82 @@ func (d *Daemon) getKubeProxyReplacementStatus() *models.KubeProxyReplacement {
+				Size: int64(ipcachemap.MaxEntries),
+			},
+			{
+				Name: "IP masquerading",
```
```suggestion
				Name: "IP masquerading agent",
```

brb commented on daemon/cmd/status.go:
```diff
@@ -200,6 +210,106 @@ func (d *Daemon) getKubeProxyReplacementStatus() *models.KubeProxyReplacement {
+				Size: int64(lbmap.MaxEntries),
+			},
+			{
+				Name: "IPv4 backend", // cilium_lb4_backends
```
```suggestion
				Name: "IPv4 service backend", // cilium_lb4_backends
```

brb commented on daemon/cmd/status.go:
```diff
@@ -200,6 +210,106 @@ func (d *Daemon) getKubeProxyReplacementStatus() *models.KubeProxyReplacement {
+				Size: int64(lbmap.MaxEntries),
+			},
+			{
+				Name: "IPv4 reverse NAT", // cilium_lb4_reverse_nat
```
```suggestion
				Name: "IPv4 service reverse NAT", // cilium_lb4_reverse_nat
```
